src/libxtp/eigencuda.cc

```diff
@@ -0,0 +1,183 @@
+/*
+ *            Copyright 2009-2019 The VOTCA Development Team
+ *                       (http://www.votca.org)
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License")
+ *
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *              http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include <votca/xtp/eigencuda.h>
+
+namespace votca {
+namespace xtp {
+
+/*
+ * Deallocate memory from the device
+ */
+void free_mem_in_gpu(double *x) { checkCuda(cudaFree(x)); };
+
+EigenCuda::~EigenCuda() {
+
+  // destroy handle
+  cublasDestroy(_handle);
+  // destroy stream
+  cudaStreamDestroy(_stream);
+}
+
+/*
+ * Check if the available memory is enough to compute the system
+ */
+void EigenCuda::check_available_memory_in_gpu(size_t requested_memory) const {
+  size_t free, total;
+  checkCuda(cudaMemGetInfo(&free, &total));
+
+  std::ostringstream oss;
+  oss << "There were requested : " << requested_memory
+      << "bytes int the device\n";
+  oss << "Device Free memory (bytes): " << free
+      << "\nDevice total Memory (bytes): " << total << "\n";
+
+  // Raise an error if there is not enough total or free memory in the device
+  if (requested_memory > free) {
+    oss << "There is not enough memory in the Device!\n";
+    throw std::runtime_error(oss.str());
+  }
+}
+
+/*
+ * Allocate memory in the device for matrix A.
+ */
+uniq_double EigenCuda::alloc_matrix_in_gpu(size_t size_matrix) const {
+
+  // Pointer in the device
+  double *dmatrix;
+  checkCuda(cudaMalloc(&dmatrix, size_matrix));
+  uniq_double dev_ptr(dmatrix, free_mem_in_gpu);
+  return dev_ptr;
+}
+
+uniq_double EigenCuda::copy_matrix_to_gpu(const Eigen::MatrixXd &matrix) const {
+  // allocate memory in the device
+  size_t size_matrix = matrix.size() * sizeof(double);
+  uniq_double dev_ptr = alloc_matrix_in_gpu(size_matrix);
+
+  // Transfer data to the GPU
```

Comment:
I feel this could all be part of CudaMatrix

Reply:
Yes, see new implementation